include/appbase/execution_priority_queue.hpp

```diff
@@ -0,0 +1,172 @@
+#pragma once
+#include <boost/asio.hpp>
+
+#include <queue>
+
+namespace appbase {
+// adapted from: https://www.boost.org/doc/libs/1_69_0/doc/html/boost_asio/example/cpp11/invocation/prioritised_handlers.cpp
+
+struct priority {
+   static constexpr int high = 100;
+   static constexpr int medium = 50;
+   static constexpr int low = 10;
+};
+
+class execution_priority_queue : public boost::asio::execution_context
+{
+public:
+
+   template <typename Function>
+   void add(int priority, Function function)
+   {
+      std::unique_ptr<queued_handler_base> handler(new queued_handler<Function>(priority, std::move(function)));
+
+      handlers_.push(std::move(handler));
+   }
+
+   void execute_all()
+   {
+      while (!handlers_.empty())
+      {
+         handlers_.top()->execute();
+         handlers_.pop();
+      }
+   }
+
+   bool execute_highest()
+   {
+      // execute at least the highest priority, and all of the available >= priority::high priority
+      while( !handlers_.empty() )
+      {
+         const auto& top = handlers_.top();
+         const int priority = top->priority();
+         top->execute();
+         handlers_.pop();
+         if( priority < priority::high ) {
+            // execute only one after executing all of high
+            break;
+         }
+      }
+
+      return !handlers_.empty();
+   }
+
+   void execute_high() {
```

Comment:
conceptually we seem to have the  competing ideas:
* priorities are continuous
* priorities are categories/buckets

From an implementation point of view, it is currently continuous and in the application level code that uses this we even sometimes treat it as continuous by doing things like `priority::low-1` 

However, that means this method and its interaction with `application::exec` violate some of my expectations as a user which is that any higher priority item that is posted/known at time T will execute before a lower priority item.

This method will execute all known actions at or above `priority::high` known at `T0` in priority order however, new jobs posted during that process will not be reflected in the priority queue until `application::exec` calls `(run|poll)_one`.

So, from the outside, I would expect to be able to atomically post:

Job A @ priority 101 
Job B @ priority 100
where Job A will post Job C @ priority 100
And see the execution order of A,C,B under the idea that "Time" moves after each job:
T0: A,B known, execute A
T1: C,B known, execute C
T2: B known, execute B

As it stands, I'm pretty sure I will see A,B,C "Time" doesn't advance until B is retired.

Reply:
I lost the original point I was trying to make, if priority was a category and all Highs were the same, then A,B,C and A,C,B would both be acceptable and this method and its interaction with `application::run` would make sense from a user perspective as all of those priority fall into the "high" category.  

If we want to do that, then we should probably reflect the broad categories in the API instead of allowing a continuous priority which behaves oddly around some point in the range.

If we want to keep priority as a continuous, then I think the simpler approach of applying all posted io-service jobs, then picking the top of the queue and repeating gives us the expected execution order and probably less code to maintain at some cost of overhead.

Reply:
I think your right, KISS is probably the best.  The overhead of posting all io-service jobs should be trivial.